### PR TITLE
Add Test Case Exclusions

### LINKED ISSUE #1701: Proposal: Method / Class / Namespace exclusions as XUnitFilters
<h2>Problem</h2>
Currently XUnit doesn't support a way of dynamically skipping methods, classes or namespaces. Disabling tests requires adding annotations to source code, which is often not feasible or convenient - e.g. when multiple projects need to run the same test suite. Considering that assemblies can be explicitly excluded from running, exclusion of methods seems like a fairly straightforward addition. 

<h2>Proposal</h2>
Mirror the implementation for included methods, namespaces and classes in `XunitFilters.cs` to allow for exclusions of the above. 

https://github.com/xunit/xunit/blob/6bd342df4ea50477ada6e128f366cb8756b3016d/src/xunit.runner.utility/Project/XunitFilters.cs#L35-L55

This would allow for Xunit to be run with a (for example) `-nomethod`  or `-noclass` option from the command line. 

<h2>Comments</h2>
The behavior of Included/excluded traits seems to specify that inclusion supersedes exclusion, so this will be mimicked here as well.
The only mention of this I could find was in #791 , which seems to have been closed for age. 

Has this feature been considered before? 
Seems like the implementation would be straight-forward and if the change is considered appropriate to the project, I would submit a PR with the changes.

### PR DESCRIPTION
**Pull request by A-And**
Fixes #1701. 

Mirror the implementation for including and excluding methods, classes and namespaces.

cc @ViktorHofer

### CODE REVIEW

File: src/dotnet-xunit/Program.cs [hunk 1]
```diff
@@ -229,11 +229,19 @@ static void PrintUsage()
...
+        Console.WriteLine("                         : if specified more than once, acts as an AND operation");
         Console.WriteLine("  -class \"name\"          : run all methods in a given test class (should be fully");
         Console.WriteLine("                         : specified; i.e., 'MyNamespace.MyClass')");
         Console.WriteLine("                         : if specified more than once, acts as an OR operation");
+        Console.WriteLine("  -noclass \"name\"        : do not run any methods in a given test class (should be fully");
+        Console.WriteLine("                         : specified; i.e., 'MyNamespace.MyClass')");
+        Console.WriteLine("                         : if specified more than once, acts as an AND operation");
         Console.WriteLine("  -namespace \"name\"      : run all methods in a given namespace (i.e.,");
         Console.WriteLine("                         : 'MyNamespace.MySubNamespace')");
+        Console.WriteLine("  -nonamespace \"name\"    : do not run any methods in a given namespace (i.e.,");
```
Comment (ViktorHofer) [inline src/dotnet-xunit/Program.cs]: L245 `Console.WriteLine("                         : if specified more than once, acts as an OR operation");` belongs to the command above "-namespace". If the OR also applies to "-nonamespace" you should probably copy and paste that line above.

File: src/xunit.console/ConsoleRunner.cs [hunk 2]
```diff
@@ -266,11 +266,19 @@ void PrintUsage(IReadOnlyList<IRunnerReporter> reporters)
...
             Console.WriteLine("                         : specified; i.e., 'MyNamespace.MyClass')");
             Console.WriteLine("                         : if specified more than once, acts as an OR operation");
+            Console.WriteLine("  -noclass \"name\"        : do not run any methods in a given test class (should be fully");
+            Console.WriteLine("                         : specified; i.e., 'MyNamespace.MyClass')");
+            Console.WriteLine("                         : if specified more than once, acts as an AND operation");
             Console.WriteLine("  -namespace \"name\"      : run all methods in a given namespace (i.e.,");
             Console.WriteLine("                         : 'MyNamespace.MySubNamespace')");
+            Console.WriteLine("  -nonamespace \"name\"    : do not run any methods in a given namespace (i.e.,");
+            Console.WriteLine("                         : 'MyNamespace.MySubNamespace')");
             Console.WriteLine("                         : if specified more than once, acts as an OR operation");
```
Comment (ViktorHofer) [inline src/xunit.console/ConsoleRunner.cs]: same here.

File: src/xunit.runner.utility/Project/XunitFilters.cs [hunk 3]
```diff
@@ -40,15 +47,30 @@ public XunitFilters()
...
+
         /// <summary>
         /// Gets the set of method filters for tests to include.
         /// </summary>
         public ICollection<string> IncludedMethods => includedMethods;
 
+        /// <summary>
+        /// Gets the set of assembly filters for tests to exclude.
+        /// </summary>
+        public HashSet<string> ExcludedNameSpaces { get; }
```
Comment (ViktorHofer) [inline src/xunit.runner.utility/Project/XunitFilters.cs]: nit: "ExcludedNamespaces", according to https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/namespaces/using-namespaces

Comment (A-And) [reply src/xunit.runner.utility/Project/XunitFilters.cs]: Here I've stuck to the convention above, i.e. `IncludedNameSpaces` .

File: src/dotnet-xunit/Program.cs (hunk 1 above)
Comment (A-And) [reply src/dotnet-xunit/Program.cs]: Good catch, thanks!

File: src/xunit.runner.utility/Project/XunitFilters.cs (hunk 3 above)
Comment (bradwilson) [reply src/xunit.runner.utility/Project/XunitFilters.cs]: I'll change it after merge.

**Comment by bradwilson** [comment]
Thanks!

**Pull request merged**